### ISSUE TITLE
Query Builder

### ISSUE THREAD

**Issue by ehsannm** [dgraph-io/dgraph#3737]
Writing queries in a text making the code too hard. We have to care about spelling and even syntax check is too hard. Do you have any query builder for dgraph ?

**Comment by MichelDiz** [comment]
We have the Ratel UI https://github.com/dgraph-io/ratel

**Comment by ehsannm** [comment]
maybe in wrong repo, i meant for example on go client driver. dgo
It has only one command for query which supports text and one with vars. but having a query builder help to use the driver better

**Comment by gitlw** [comment]
Hi @ehsannm ,  for now we don't have a plan to support the query builder as our query language is fairly flexible, and creating the query builder would probably generate too many states in the client.

To me, a fair comparison is that you normally don't have a query builder when writing a SQL statement, since SQL is also a quite flexible query language.

**Comment by nicklanng** [comment]
I  started on a Go query builder this weekend: https://github.com/nicklanng/plusminus

**Comment by glacierck** [comment]
implemented using Python please!!!!!!!!😭😭

**Comment by dan-j** [comment]
> To me, a fair comparison is that you normally don't have a query builder when writing a SQL statement, since SQL is also a quite flexible query language.

I don't necessarily agree with this statement:

- There are numerous SQL ORM libraries for every popular language I can think of which provide query builders
- https://github.com/acro5piano/typed-graphqlify for generating type-safe GraphQL queries/mutations in TypeScript
- https://github.com/MichalLytek/type-graphql for type-safe schema definition in TypeScript.

Not sure how you'd implement them in Go, but I think it's worth more thought?

**Comment by binier** [comment]
Hi. Wrote [query builder in TypeScript](https://github.com/binier/dgraphium). It's still rather new but usable.

**Comment by Mentioum** [comment]
For others who come here thinking about this issue I ended up writing a code generator which generated all queries and mutatations for CRUD including support for individual field updating on nodes. (In Go)

I did it using https://github.com/valyala/fasttemplate and the DQL schema definitions.  Its very internal code atm (works specifically for us) but I would consider making an OS version if there enough interest.  Not sure if there would be though given all the recent rush to GraphQL on Slash.
